src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -0,0 +1,277 @@
+        if (formattingScope == FormattingScope.CODE_BLOCK) {
+            int offset = editor.getCaretModel().getOffset();
+            int beforeOffset = offset - 1;
+            TextRange codeBlockRange = LSPCodeBlockProvider.getCodeBlockRange(editor, file, beforeOffset);
+            if (codeBlockRange != null) {
+                int startOffset = codeBlockRange.getStartOffset();
+                int endOffset = codeBlockRange.getEndOffset();
+
+                // Make sure the range includes the brace pair
+                Document document = editor.getDocument();
+                CharSequence documentChars = document.getCharsSequence();
+                if ((startOffset > 0) && (documentChars.charAt(startOffset) != openBraceChar)) {
+                    startOffset--;
+                }
+                if ((endOffset < (documentChars.length() - 1)) && (documentChars.charAt(endOffset) != closeBraceChar)) {
+                    endOffset++;
+                }
+
+                // If the range is now the braced block, format it
+                if ((documentChars.charAt(startOffset) == openBraceChar) && (documentChars.charAt(endOffset) == closeBraceChar)) {
+                    formatTextRange = TextRange.create(startOffset, endOffset);
+                }
+            }
+        }
+
+        // If appropriate, use the file text range
+        else if (formattingScope == FormattingScope.FILE) {
+            formatTextRange = file.getTextRange();
+        }
+
+        // If we have a text range now, format it
+        if (formatTextRange != null) {
+            CodeStyleManager.getInstance(project).reformatText(file, Collections.singletonList(formatTextRange));
+            return Result.STOP;
+        }
+
+        return Result.CONTINUE;
+    }
+
+    @NotNull
+    private static Result handleStatementTerminatorTyped(@NotNull Project project,
+                                                         @NotNull Editor editor,
+                                                         @NotNull PsiFile file,
+                                                         @NotNull LSPFormattingFeature formattingFeature,
+                                                         char statementTerminatorChar) {
+        TextRange formatTextRange = null;
+
+        int offset = editor.getCaretModel().getOffset();
+        int beforeOffset = offset - 1;
+
+        // If appropriate, find the statement that was just terminated
+        FormattingScope formattingScope = formattingFeature.getFormatOnStatementTerminatorScope(file);
+        if (formattingScope == FormattingScope.STATEMENT) {
+            List<TextRange> selectionTextRanges = LSPSelectionRangeSupport.getSelectionTextRanges(file, editor, beforeOffset);
+            if (!ContainerUtil.isEmpty(selectionTextRanges)) {
+                // Find the closest selection range that is extended to line start/end; that should be the statement
```

Comment:
Because we don't have access to an actual AST/PSI tree here, we have to use a bit of a heuristic to find the statement. Language server support for selection ranges is almost always directly bound to the underlying AST, and therefore one of the selection ranges shouuld be for a statement in that language. And that selection range will be for full lines where extension of range start to line start and range end to line end is already performed by `LSPSelectionRangeSupport.getSelectionTextRanges()`. So the selection range that represents a contiguous range of full lines should in fact be the entire statement. Let me know if that doesn't make sense because it's the foundation upon which statement scope operates.